### PR TITLE
Deprecate `project.buildDir`

### PR DESCRIPTION
**Pull request by ljacomet**

### CODE REVIEW

**Comment by jvandort** [comment]
Will we be doing the same for `Project#getProjectDir`?

**Comment by ljacomet** [comment]
Certainly not in this PR. But I am not sure generally. That property is read-only and so less of an issue.

**Comment by ljacomet** [comment]
@bot-gradle test this

**Comment by ljacomet** [comment]
@bot-gradle test this

**Comment by ljacomet** [comment]
@bot-gradle test this

**Comment by jvandort** [comment]
A related question, why is the build dir location configurable at all? Are there use cases for changing it?

**Comment by ljacomet** [comment]
Yes, there are use cases. One is not having any build produced files inside your project source tree.

**Comment by ljacomet** [comment]
@bot-gradle test this

**Comment by ljacomet** [comment]
@bot-gradle test this

**Comment by ljacomet** [comment]
@bot-gradle test RfR

**Comment by ljacomet** [comment]
Looks like RfR is really unstable for now.

Let's see if this would pass nightly at least

**Comment by ljacomet** [comment]
@bot-gradle test RfN

**Comment by ljacomet** [comment]
@bot-gradle test RfN

**Review by ov7a — CHANGES_REQUESTED** [review]
While code changes are ok, I think this change should be reflected in the release notes and documentation with a detailed explanation of the reasons behind this change.

File: subprojects/kotlin-dsl/src/main/kotlin/org/gradle/kotlin/dsl/support/delegates/ProjectDelegate.kt [hunk 1]
```diff
@@ -103,6 +103,8 @@ abstract class ProjectDelegate : Project {
     override fun getLogger(): Logger =
         delegate.logger
 
+    @Suppress("DEPRECATION")
+    @Deprecated("Use layout.buildDirectory instead", ReplaceWith("layout.buildDirectory.get()"))
     override fun getBuildDir(): File =
         delegate.buildDir
```
Comment (eskatos) [inline subprojects/kotlin-dsl/src/main/kotlin/org/gradle/kotlin/dsl/support/delegates/ProjectDelegate.kt:109]: ❔ Can't the deprecation suppression be reduced to the call to the delegate as in `setBuildDir` below?

```suggestion
    @Deprecated("Use layout.buildDirectory instead", ReplaceWith("layout.buildDirectory.get()"))
    override fun getBuildDir(): File =
        @Suppress("DEPRECATION")
        delegate.buildDir
```

File: subprojects/kotlin-dsl/src/main/kotlin/org/gradle/kotlin/dsl/support/delegates/ProjectDelegate.kt [hunk 2]
```diff
@@ -244,10 +246,14 @@ abstract class ProjectDelegate : Project {
...
+    @Deprecated("Use layout.buildDirectory instead", ReplaceWith("layout.buildDirectory.set(path)"))
     override fun setBuildDir(path: File) {
+        @Suppress("DEPRECATION")
         delegate.buildDir = path
     }
 
+    @Suppress("DEPRECATION")
+    @Deprecated("Use layout.buildDirectory instead", ReplaceWith("layout.buildDirectory.set(file(path))"))
     override fun setBuildDir(path: Any) =
         delegate.setBuildDir(path)
```
Comment (eskatos) [inline subprojects/kotlin-dsl/src/main/kotlin/org/gradle/kotlin/dsl/support/delegates/ProjectDelegate.kt:258]: ❔ Can't the deprecation suppression be reduced to the call to the delegate as in `setBuildDir` above?


```suggestion
    @Deprecated("Use layout.buildDirectory instead", ReplaceWith("layout.buildDirectory.get()"))
    override fun setBuildDir(path: Any) =
        @Suppress("DEPRECATION")
        delegate.setBuildDir(path)
```

**Review by eskatos — APPROVED** [review]

**Comment by ljacomet** [comment]
@ov7a I'll work on documenting these changes in a follow up PR. Note that as indicated on the issue, we will not start nagging users now.

**Comment by ljacomet** [comment]
@bot-gradle test and merge

**Review by lkasso — APPROVED** [review]

**Review by ov7a — APPROVED** [review]

**Pull request merged**